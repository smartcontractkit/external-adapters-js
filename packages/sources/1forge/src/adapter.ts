import { Builder, Logger, Requester, Validator } from '@chainlink/ea-bootstrap'
import {
  DefaultConfig,
  ExecuteWithConfig,
  ExecuteFactory,
  AdapterRequest,
  APIEndpoint,
  MakeWSHandler,
} from '@chainlink/ea-bootstrap'
import { DEFAULT_WS_API_ENDPOINT, makeConfig } from './config'
import * as endpoints from './endpoint'

export const execute: ExecuteWithConfig<DefaultConfig, endpoints.TInputParameters> = async (
  request,
  context,
  config,
) => {
  return Builder.buildSelector<DefaultConfig, endpoints.TInputParameters>(
    request,
    context,
    config,
    endpoints,
  )
}

export const endpointSelector = (
  request: AdapterRequest,
): APIEndpoint<DefaultConfig, endpoints.TInputParameters> =>
  Builder.selectEndpoint<DefaultConfig, endpoints.TInputParameters>(
    request,
    makeConfig(),
    endpoints,
  )

export const makeExecute: ExecuteFactory<DefaultConfig, endpoints.TInputParameters> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}

export const makeWSHandler = (config?: DefaultConfig): MakeWSHandler => {
  const getSubscription = (symbol?: string, subscribe = true): string => {
    if (!symbol) return ''
    return subscribe ? `subscribe_to|${symbol}` : `unsubscribe_from|${symbol}`
  }
  const getSymbol = (input: AdapterRequest) => {
    const validator = new Validator(
      input,
      endpoints.quotes.inputParameters,
      {},
      { shouldThrowError: false },
    )
    if (validator.error) return
    if (Array.isArray(validator.validated.data.quote)) {
      Logger.debug(
        `[WS]: ${validator.validated.data.quote} supplied as quote. Only non-array tickers can be used for WS`,
      )
      return
    }
    if (Array.isArray(validator.validated.data.base)) {
      Logger.debug(
        `[WS]: ${validator.validated.data.base} supplied as base. Only non-array tickers can be used for WS`,
      )
      return
    }
    const symbol = validator.validated.data.base.toUpperCase()
    const convert = validator.validated.data.quote.toUpperCase()
    return `${symbol}/${convert}`
  }
  const parseResponse = (response: string) => {
    const [type, data] = response.split('|')
    return { type, data: JSON.parse(data) }
  }

  return () => {
    const defaultConfig = config || makeConfig()

    return {
      connection: {
        url: DEFAULT_WS_API_ENDPOINT,
      },
      subscribe: (input) => getSubscription(getSymbol(input)),
      unsubscribe: (input) => getSubscription(getSymbol(input), false),
      subsFromMessage: (message: any) => {
        if (!message.data || message.data.indexOf('update') === -1) return
        const { data } = parseResponse(message.data)
        return getSubscription(data.s)
      },
      isError: () => false,
      filter: (message: any) => message.data.indexOf('update') !== -1,
      toResponse: (message: any) => {
        const { data } = parseResponse(message.data)
        return Requester.success('1', { data: { result: data.p } })
      },
      onConnectChain: [
        {
          payload: `login|${defaultConfig.apiKey}`,
          filter: (message: any) => message.data.indexOf('post_login_success') !== -1,
        },
      ],
    }
  }
}
