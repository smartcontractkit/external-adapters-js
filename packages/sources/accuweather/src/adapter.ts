import { Builder } from '@chainlink/ea-bootstrap'
import {
  DefaultConfig,
  ExecuteWithConfig,
  ExecuteFactory,
  AdapterRequest,
  APIEndpoint,
} from '@chainlink/ea-bootstrap'
import { makeConfig } from './config'
import * as endpoints from './endpoint'

export const execute: ExecuteWithConfig<DefaultConfig, endpoints.TInputParameters> = async (
  request,
  context,
  config,
) => {
  return Builder.buildSelector<DefaultConfig, endpoints.TInputParameters>(
    request,
    context,
    config,
    endpoints,
  )
}

export const endpointSelector = (
  request: AdapterRequest,
): APIEndpoint<DefaultConfig, endpoints.TInputParameters> =>
  Builder.selectEndpoint<DefaultConfig, endpoints.TInputParameters>(
    request,
    makeConfig(),
    endpoints,
  )

export const makeExecute: ExecuteFactory<DefaultConfig, endpoints.TInputParameters> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}
