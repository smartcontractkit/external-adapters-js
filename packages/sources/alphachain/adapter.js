const { Requester, Validator } = require('@chainlink/external-adapter')
const { util } = require('@chainlink/ea-bootstrap')

const customError = (data) => {
  if (data.status !== '200') return true
  return false
}

const customParams = {
  base: ['base', 'from', 'coin'],
  quote: ['quote', 'to', 'market'],
}

const execute = (input, callback) => {
  const validator = new Validator(input, customParams)
  if (validator.error) return callback(validator.error.statusCode, validator.errored)

  const jobRunID = validator.validated.id
  const url = 'https://alpha-chain2.p.rapidapi.com/data-query'
  const host = 'alpha-chain2.p.rapidapi.com'
  const headers = {
    'content-type': 'application/octet-stream',
    'x-rapidapi-host': host,
    'x-rapidapi-key': util.getRandomRequiredEnv('API_KEY'),
    useQueryString: true,
  }
  const base = validator.validated.data.base.toUpperCase()
  const quote = validator.validated.data.quote.toUpperCase()

  const params = {
    from_symbol: base,
    to_symbol: quote,
    chainlink_node: true,
  }

  const config = {
    url,
    params,
    headers,
  }

  Requester.request(config, customError)
    .then((response) => {
      response.data.result = Requester.validateResultNumber(response.data, ['result'])
      callback(response.status, Requester.success(jobRunID, response))
    })
    .catch((error) => callback(500, Requester.errored(jobRunID, error)))
}

module.exports.execute = execute
