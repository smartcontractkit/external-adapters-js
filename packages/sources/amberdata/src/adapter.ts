import { Requester, Validator, Builder } from '@chainlink/ea-bootstrap'
import {
  AdapterRequest,
  Config,
  ExecuteFactory,
  ExecuteWithConfig,
  MakeWSHandler,
  APIEndpoint,
} from '@chainlink/types'
import { DEFAULT_WS_API_ENDPOINT, makeConfig, NAME } from './config'
import * as endpoints from './endpoint'
import { crypto } from './endpoint'

// Export function to integrate with Chainlink node
export const execute: ExecuteWithConfig<Config> = async (request, context, config) => {
  return Builder.buildSelector(request, context, config, endpoints)
}

export const endpointSelector = (request: AdapterRequest): APIEndpoint =>
  Builder.selectEndpoint(request, makeConfig(), endpoints)

export const makeExecute: ExecuteFactory<Config> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}

export const makeWSHandler = (defaultConfig?: Config): MakeWSHandler => {
  const subscriptions: any = {}
  const getPair = (input: AdapterRequest) => {
    const validator = new Validator(input, crypto.inputParameters, {}, false)
    if (validator.error) return
    const base = (validator.overrideSymbol(NAME) as string).toLowerCase()
    const quote = validator.validated.data.quote.toLowerCase()
    return `${base}_${quote}`
  }
  const getSubscription = (pair?: string) => {
    if (!pair) return
    return { id: 1, method: 'subscribe', params: ['market:tickers', { pair }] }
  }
  const getUnsubscription = (pair?: string) => {
    if (!pair) return
    return { id: 1, method: 'unsubscribe', params: [subscriptions[pair]] }
  }
  return () => {
    const config = defaultConfig || makeConfig()
    return {
      connection: {
        url: config.api.baseWsURL || DEFAULT_WS_API_ENDPOINT,
        protocol: { headers: { ...config.api.headers } },
      },
      subscribe: (input) => getSubscription(getPair(input)),
      unsubscribe: (input) => getUnsubscription(getPair(input)),
      subsFromMessage: (message) => {
        const pair = message?.params?.result?.pair
        subscriptions[pair] = message?.params?.subscription
        return getSubscription(message?.params?.result?.pair)
      },
      // https://github.com/web3data/web3data-js/blob/5b177803cb168dcaed0a8a6e2b2fbd835b82e0f9/src/websocket.js#L43
      isError: () => false, // Amberdata never receives error types?
      filter: (message: any) => !!message.params,
      toResponse: (message: any) => {
        const result = Requester.validateResultNumber(message, ['params', 'result', 'last'])
        return Requester.success('1', { data: { result } })
      },
    }
  }
}
