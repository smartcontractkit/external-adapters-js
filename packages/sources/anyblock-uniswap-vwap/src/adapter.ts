import { Execute } from '@chainlink/types'
import { Requester, Validator } from '@chainlink/ea-bootstrap'
import { util } from '@chainlink/ea-bootstrap'

const customError = (data: any) => {
  return !data.hits || !data.hits.hits || data.hits.hits.length < 1
}

const customParams = {
  address: true,
  debug: false,
  roundDay: false,
  start: false,
  end: false,
}

const buildVWAP = (response: any, debug: boolean) => {
  const sources = response.data.hits.hits.map((i: any) => {
    const reserve0 = i._source.args.find((j: any) => j.pos === 0)
    const reserve1 = i._source.args.find((j: any) => j.pos === 1)
    return {
      timestamp: i._source.timestamp,
      reserve0: parseInt(reserve0['value.hex'], 16),
      reserve1: parseInt(reserve1['value.hex'], 16),
    }
  })

  let overallVolume = 0
  let sumAmountAndPrices = 0
  for (let i = 1; i < sources.length; i++) {
    const reserve0volume = Math.abs(sources[i].reserve0 - sources[i - 1].reserve0)
    const price = sources[i].reserve0 / sources[i].reserve1
    overallVolume += reserve0volume
    sumAmountAndPrices += price * reserve0volume
  }

  const vwap = sumAmountAndPrices / overallVolume
  const resp: any = {
    status: response.status,
    data: { result: vwap },
  }

  if (debug) resp.data.raw = sources
  return resp
}

const cleanupDate = (inputDate: string, roundDay: boolean) => {
  let outputDate: number
  try {
    outputDate = parseInt(inputDate)
    if (roundDay) {
      const date = new Date(outputDate)
      date.setUTCHours(0, 0, 0, 0)
      outputDate = date.getTime()
    }
  } catch (err) {
    return inputDate
  }
  return outputDate
}

// TODO: enable other networks
const getAnyblockUrl = () => 'https://api.anyblock.tools/ethereum/ethereum/mainnet/es/event/search/'

export const execute: Execute = async (input) => {
  const validator = new Validator(input, customParams)
  if (validator.error) throw validator.error

  const jobRunID = validator.validated.id
  // TODO: validate this is a checksum address
  const address = validator.validated.data.address
  const debug = validator.validated.data.debug || false
  const roundDay = validator.validated.data.roundDay || false
  let start = validator.validated.data.start
  let end = validator.validated.data.end

  const url = getAnyblockUrl()

  end = cleanupDate(end, roundDay)
  start = cleanupDate(start, roundDay)

  if (!start && !end) {
    const date = new Date()
    date.setUTCHours(0, 0, 0, 0)
    end = date.getTime() / 1000

    start = end - 60 * 60 * 24
  } else if (!start) {
    start = end - 60 * 60 * 24
  } else if (!end) {
    end = start + 60 * 60 * 24
  } else if (start === end) {
    start = end - 60 * 60 * 24
  }

  const body = {
    query: {
      bool: {
        filter: [
          { term: { 'address.raw': address } },
          { term: { 'event.raw': 'Sync' } },
          {
            range: {
              timestamp: {
                gte: start,
                lte: end,
              },
            },
          },
        ],
      },
    },
    sort: [{ timestamp: 'asc' }],
    size: 10000,
    _source: ['timestamp', 'args'],
  }

  const headers = {
    'Content-Type': 'application/json',
    Authorization: `Bearer ${util.getRandomRequiredEnv('API_KEY')}`,
  }

  const config = {
    url,
    headers,
    data: body,
  }

  const response = await Requester.request(config, customError)
  const vwapResp = buildVWAP(response, debug)
  return Requester.success(jobRunID, vwapResp)
}
