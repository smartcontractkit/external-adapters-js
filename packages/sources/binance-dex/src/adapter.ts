import { Requester } from '@chainlink/ea-bootstrap'
import { Config, ExecuteWithConfig, ExecuteFactory } from '@chainlink/types'
import { makeConfig } from './config'
import { price } from './endpoint'

export const execute: ExecuteWithConfig<Config> = async (request, context, config) => {
  Requester.logConfig(config)
  return await price.execute(request, context, config)
}

export const makeExecute: ExecuteFactory<Config> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}
