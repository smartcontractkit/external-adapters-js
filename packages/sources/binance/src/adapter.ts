import { Requester, Validator, AdapterError } from '@chainlink/ea-bootstrap'
import { Config, ExecuteWithConfig, ExecuteFactory, MakeWSHandler, AdapterRequest } from '@chainlink/types'
import { makeConfig, DEFAULT_ENDPOINT, DEFAULT_WS_API_ENDPOINT } from './config'
import { ticker } from './endpoint'

const inputParams = {
  endpoint: false,
}

export const execute: ExecuteWithConfig<Config> = async (request, config) => {
  const validator = new Validator(request, inputParams)
  if (validator.error) throw validator.error

  Requester.logConfig(config)

  const jobRunID = validator.validated.id
  const endpoint = validator.validated.data.endpoint || DEFAULT_ENDPOINT

  switch (endpoint.toLowerCase()) {
    case ticker.NAME: {
      return await ticker.execute(request, config)
    }
    default: {
      throw new AdapterError({
        jobRunID,
        message: `Endpoint ${endpoint} not supported.`,
        statusCode: 400,
      })
    }
  }
}

export const makeExecute: ExecuteFactory<Config> = (config) => {
  return async (request) => execute(request, config || makeConfig())
}

export const makeWSHandler = (config?: Config): MakeWSHandler => {
  const getSubscription = (symbol?: string, subscribe = true) => {
    if (!symbol) return
    return {
      "method": subscribe ? 'SUBSCRIBE' : 'UNSUBSCRIBE',
      "params":
        [
          `${symbol}@miniTicker`,
        ],
      "id": 1
    }
  }
  const getSymbol = (input: AdapterRequest) => {
    const validator = new Validator(input, ticker.customParams, {}, false)
    if (validator.error) return
    const symbol = validator.validated.data.base.toUpperCase()
    const convert = validator.validated.data.quote.toUpperCase()
    return `${symbol.toLowerCase()}${convert.toLowerCase()}`
  }
  return () => {
    const defaultConfig = config || makeConfig()
    return {
      connection: {
        url: defaultConfig.api.baseWsURL || DEFAULT_WS_API_ENDPOINT,
      },
      subscribe: (input) => getSubscription(getSymbol(input)),
      unsubscribe: (input) => getSubscription(getSymbol(input), false),
      subsFromMessage: (message) => {
        if (!message.s) return undefined
        return getSubscription(`${message.s.toLowerCase()}`)
      },
      isError: (message: any) => message.type === 'error',
      // Ignore everything is not a ticker message. Throw an error on incoming errors.
      filter: (message: any) => message.e === '24hrMiniTicker',
      toResponse: (message: any) => {
        const result = Requester.validateResultNumber(message, ['c'])
        return Requester.success('1', { data: { result } })
      },
    }
  }
}
