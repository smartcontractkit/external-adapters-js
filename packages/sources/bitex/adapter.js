const { Requester, Validator } = require('@chainlink/external-adapter')

const customParams = {
  base: ['base', 'from', 'coin'],
  quote: ['quote', 'to', 'market'],
  endpoint: false,
}

const execute = (input, callback) => {
  const validator = new Validator(input, customParams)
  if (validator.error) return callback(validator.error.statusCode, validator.errored)

  const jobRunID = validator.validated.id
  const endpoint = validator.validated.data.endpoint || 'tickers'
  const base = validator.validated.data.base.toLowerCase()
  const quote = validator.validated.data.quote.toLowerCase()
  const url = `https://bitex.la/api/${endpoint}/${base}_${quote}`

  const config = {
    url,
  }

  Requester.request(config)
    .then((response) => {
      response.data.result = Requester.validateResultNumber(response.data, [
        'data',
        'attributes',
        'vwap',
      ])
      callback(response.status, Requester.success(jobRunID, response))
    })
    .catch((error) => callback(500, Requester.errored(jobRunID, error)))
}

module.exports.execute = execute
