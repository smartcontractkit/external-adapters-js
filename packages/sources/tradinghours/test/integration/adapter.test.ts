import process from 'process'

import { ServerInstance } from '@chainlink/external-adapter-framework'
import { MarketStatus } from '@chainlink/external-adapter-framework/adapter'
import { SuperTest, Test } from 'supertest'

import { mockResponseSuccess } from './fixtures'
import { setupExternalAdapterTest, SuiteContext } from './setup'

describe('Market status endpoint', () => {
  let spy: jest.SpyInstance

  beforeAll(async () => {
    const mockDate = new Date('2022-01-01T11:11:11.111Z')
    spy = jest.spyOn(Date, 'now').mockReturnValue(mockDate.getTime())
  })

  afterAll((done) => {
    spy.mockRestore()
    done()
  })

  const context: SuiteContext = {
    req: null,
    server: async () => {
      // workaround for failing integration tests that run in parallel
      process.env['RATE_LIMIT_CAPACITY_SECOND'] = '10000'
      process.env['METRICS_ENABLED'] = 'false'
      const server = (await import('../../src')).server
      return server() as Promise<ServerInstance>
    },
  }

  const envVariables = {
    API_KEY: 'fake-api-key',
  }

  setupExternalAdapterTest(envVariables, context)

  const openData = {
    data: {
      endpoint: 'market-status',
      market: 'forex',
    },
  }
  const closedData = {
    data: {
      endpoint: 'market-status',
      market: 'metals',
    },
  }

  it('should return success with open', async () => {
    mockResponseSuccess()

    const response = await (context.req as SuperTest<Test>)
      .post('/')
      .send(openData)
      .set('Accept', '*/*')
      .set('Content-Type', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200)
    expect(response.body).toMatchSnapshot()
    expect(response.body.result).toEqual(MarketStatus.OPEN)
  })

  it('should return success with closed', async () => {
    mockResponseSuccess()

    const response = await (context.req as SuperTest<Test>)
      .post('/')
      .send(closedData)
      .set('Accept', '*/*')
      .set('Content-Type', 'application/json')
      .expect('Content-Type', /json/)
      .expect(200)
    expect(response.body).toMatchSnapshot()
    expect(response.body.result).toEqual(MarketStatus.CLOSED)
  })
})
