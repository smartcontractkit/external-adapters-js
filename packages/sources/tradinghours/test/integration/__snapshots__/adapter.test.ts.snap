// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Market status endpoint should return success with closed 1`] = `
{
  "data": {
    "result": 1,
  },
  "result": 1,
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 1641035471111,
    "providerDataRequestedUnixMs": 1641035471111,
    "providerIndicatedTimeUnixMs": 1717191454000,
  },
}
`;

exports[`Market status endpoint should return success with open 1`] = `
{
  "data": {
    "result": 2,
  },
  "result": 2,
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 1641035471111,
    "providerDataRequestedUnixMs": 1641035471111,
    "providerIndicatedTimeUnixMs": 1717191454000,
  },
}
`;
